iothub/service/src/Twin/TwinClient.cs

```diff
@@ -0,0 +1,744 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Common.Exceptions;
+using Microsoft.Azure.Devices.Http2;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// Subclient of <see cref="IotHubServiceClient"/> that handles getting, updating and replacing device twin and module twin.
+    /// </summary>
+    /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-csharp-csharp-twin-getstarted"/>
+    /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-csharp-csharp-module-twin-getstarted"/>
+    public class TwinClient
+    {
+        private readonly string _hostName;
+        private readonly IotHubConnectionProperties _credentialProvider;
+        private readonly HttpClient _httpClient;
+        private readonly HttpRequestMessageFactory _httpRequestMessageFactory;
+
+        private const string RequestUriFormat = "/devices/{0}";
+        private const string TwinUriFormat = "/twins/{0}";
+        private const string ModuleTwinUriFormat = "/twins/{0}/modules/{1}";
+        private static readonly TimeSpan s_regexTimeoutMilliseconds = TimeSpan.FromMilliseconds(500);
+
+        private static readonly Regex s_deviceIdRegex = new Regex(
+            @"^[A-Za-z0-9\-:.+%_#*?!(),=@;$']{1,128}$",
+            RegexOptions.Compiled | RegexOptions.IgnoreCase,
+            s_regexTimeoutMilliseconds);
+
+        // HttpMethod does not define PATCH in its enum in .netstandard 2.0, so this is the only way to create an HTTP patch request.
+        private readonly HttpMethod _patch = new HttpMethod("PATCH");
+
+        /// <summary>
+        /// Creates an instance of this class. Provided for unit testing purposes only.
+        /// </summary>
+        protected TwinClient()
+        {
+        }
+
+        internal TwinClient(string hostName, IotHubConnectionProperties credentialProvider, HttpClient httpClient, HttpRequestMessageFactory httpRequestMessageFactory)
+        {
+            _hostName = hostName;
+            _credentialProvider = credentialProvider;
+            _httpClient = httpClient;
+            _httpRequestMessageFactory = httpRequestMessageFactory;
+        }
+
+        /// <summary>
+        /// Gets <see cref="Twin"/> from IotHub
+        /// </summary>
+        /// <param name="deviceId">The device Id.</param>
+        /// <param name="cancellationToken">Task cancellation token.</param>
+        /// <returns>Twin instance.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the provided device Id is null.</exception>
```

Comment:
We can use the `<paramref>` tag here.

```suggestion
        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="deviceId"/> is null.</exception>
```

Reply:
Also elsewhere.